Test docs refactoring (#45515)
dotnet/docs: docs/core/testing/unit-testing-with-dotnet-test.md

---
title: Testing with dotnet test
description: Learn more about how dotnet test works and its support for VSTest and Microsoft.Testing.Platform (MTP)
author: Youssef1313
ms.author: ygerges
ms.date: 03/26/2025
---

# Testing with dotnet test

> In this article, you'll learn more about `dotnet test` CLI command, some history, and its support for VSTest and Microsoft.Testing.Platform (MTP).

Reviewer: rephrasing suggestion:
This article will provide you with insights into the `dotnet test` CLI command, its history and its compatibility with both VSTest and Microsoft.Testing.Platform (MTP).

Author: ```suggestion
This article will provide you with insights into the `dotnet test` CLI command, its history and its compatibility with both VSTest and Microsoft.Testing.Platform (MTP).
```

> The `dotnet test` command has two main modes of operations:

Reviewer: rephrasing suggestion:
The `dotnet test` command operates in two primary modes:

VSTest mode: This is the default mode for `dotnet test` and was the only mode available before the .NET 10 SDK. It is primarily designed for VSTest but can also run Microsoft.Testing.Platform test via [Microsoft.Testing.Platform.MSBuild](https://www.nuget.org/packages/Microsoft.Testing.Platform.MSBuild/) NuGet package.
Microsoft.Testing.Platform mode: Introduced with the .NET 10 SDK, this mode exclusively supports test applications built with Microsoft.Testing.Platform.

Author: ```suggestion
The `dotnet test` command operates in two primary modes:
```

> - VSTest mode: This is the default mode of `dotnet test`, and it's the only mode available before .NET 10 SDK. It's designed towards VSTest, but it can also run Microsoft.Testing.Platform via [Microsoft.Testing.Platform.MSBuild](https://www.nuget.org/packages/Microsoft.Testing.Platform.MSBuild/) NuGet package.
> - Microsoft.Testing.Platform mode: This is available starting with .NET 10 SDK, and only supports Microsoft.Testing.Platform test applications.

Author: ```suggestion
- VSTest mode: This is the default mode for `dotnet test` and was the only mode available before the .NET 10 SDK. It is primarily designed for VSTest but can also run Microsoft.Testing.Platform test via [Microsoft.Testing.Platform.MSBuild](https://www.nuget.org/packages/Microsoft.Testing.Platform.MSBuild/) NuGet package.
- Microsoft.Testing.Platform mode: Introduced with the .NET 10 SDK, this mode exclusively supports test applications built with Microsoft.Testing.Platform.
```

> [!TIP]
> For CLI reference, see [dotnet test](../tools/dotnet-test.md).

## VSTest mode of `dotnet test`

> For too long, VSTest has been the only test platform in .NET. As a result, `dotnet test` was designed only for VSTest. All the command-line options are VSTest oriented.

Reviewer: rephrasing suggestion:
For an extended period, VSTest has been the only test platform in .NET. Consequently, `dotnet test` was exclusively designed for VSTest, with all command-line options tailored to VSTest.

Author: ```suggestion
For an extended period, VSTest has been the only test platform in .NET. Consequently, dotnet test was exclusively designed for VSTest, with all command-line options tailored to VSTest.
```

> It works by invoking `VSTest` MSBuild target, which causes other internal targets to run and eventually runs vstest.console, translating all the `dotnet test` command-line options to equivalent in vstest.console.

Reviewer: rephrasing suggestion:
The process involves invoking the `VSTest` MSBuild target, which triggers other internal targets to run and ultimately runs vstest.console. This translates all `dotnet test` command-line options to their equivalents in vstest.console.

Author: ```suggestion
The process involves invoking the `VSTest` MSBuild target, which triggers other internal targets to run and ultimately runs vstest.console. This translates all `dotnet test` command-line options to their equivalents in vstest.console.
```

### Running MTP projects with VSTest mode

> `dotnet test` is normally capable to run VSTest projects in VSTest mode as that's what it was originally designed for. However, running MTP projects with VSTest mode is done via [Microsoft.Testing.Platform.MSBuild](https://www.nuget.org/packages/Microsoft.Testing.Platform.MSBuild). From user-end, the support is enabled by setting `TestingPlatformDotnetTestSupport` MSBuild property to true (it's false by default, for backward compatibility concerns). In simple words, setting this property to true will cause Microsoft.Testing.Platform.MSBuild to change the VSTest target behavior, and instead route it to call `InvokeTestingPlatform`, which is an MSBuild target shipped in Microsoft.Testing.Platform.MSBuild and is responsible to run MTP test applications the correct way (as executable).

Reviewer: ```suggestion
`dotnet test` is typically designed to run VSTest projects in VSTest mode, as that was its original purpose. However, to run MTP projects in VSTest mode, you can use the [Microsoft.Testing.Platform.MSBuild](https://www.nuget.org/packages/Microsoft.Testing.Platform.MSBuild). From the user's perspective, this support is enabled by setting the `TestingPlatformDotnetTestSupport` MSBuild property to true (it is false by default for backward compatibility reasons). In simple terms, setting this property to true will cause Microsoft.Testing.Platform.MSBuild to change the VSTest target behavior, redirecting it to call `InvokeTestingPlatform`. This is an MSBuild target included in Microsoft.Testing.Platform.MSBuild, responsible for correctly running MTP test applications as executables.
```

> This means that VSTest-spcific command-line options are silently ignored in this mode. For example, `--logger`.

Reviewer: ```suggestion
This means that VSTest-specific command-line options are silently ignored in this mode, such as `--logger`.
```

> This also means there should be a way to pass MTP-specific command-line options. For example, `--report-trx` (which is the equivalent of doing `--logger trx` in VSTest). As we are limited by what the `dotnet test` CLI currently has, the only way to get MTP-specific arguments is to add them after extra `--`. For example, `dotnet test -- --report-trx`.

Reviewer: ```suggestion
This implies that there should be a way to pass MTP-specific command-line options, such as `--report-trx`, which is equivalent to using `--logger trx` in VSTest. Given the current limitations of the `dotnet test` CLI, the only way to include MTP-specific arguments is by appending them after an additional `--`. For instance, `dotnet test -- --report-trx`.
```

#### Advanced technical details

> The `--` in `dotnet test` VSTest mode is used to denote RunSettings arguments. The way `dotnet test` was originally designed is to forward those RunSettings arguments as an MSBuild property named `VSTestCLIRunSettings`. So, when running MTP test applications in VSTest mode, we are using the value of `VSTestCLIRunSettings` for another purpose to denote "application arguments".

Reviewer: ```suggestion
In `dotnet test` VSTest mode, the `--` is used to indicate the RunSettings arguments. Originally, `dotnet test` was designed to pass those arguments as an MSBuild property called `VSTestCLIRunSettings`. Therefore, when running MTP test applications in VSTest mode, we repurpose the value of `VSTestCLIRunSettings` to represent the "application arguments".
```

> #### Mixing VSTest and MTP

Reviewer: ```suggestion
#### Mixing VSTest and "Microsoft.Testing.Platform (MTP)
```

> When running `dotnet test` in VSTest mode, we advise against having both VSTest and Microsoft.Testing.Platform in the same solution.

Reviewer: ```suggestion
When running `dotnet test` in VSTest mode, it is recommended to avoid including both VSTest and Microsoft.Testing.Platform in the same solution.
```

> This scenario is not officially supported, and you should pay attention to:

Reviewer: ```suggestion
This scenario is not officially supported, and you should be aware of the following:
```

> 1. VSTest-specific command-line options will only be considered for VSTest projects but not MTP test applications.

Reviewer: ```suggestion
1. VSTest-specific command-line options will only apply to VSTest projects and not to MTP test applications.
```

> 2. MTP-specific command-line options after `--` will be considered as RunSettings arguments for VSTest projects.

Reviewer: ```suggestion
2. MTP-specific command-line options provided after `--` will be treated as RunSettings arguments for VSTest projects.
```

#### Key takeaways

> 1. To run MTP test applications in `dotnet test` VSTest mode, you should be using `Microsoft.Testing.Platform.MSBuild`, be passing MTP-specific command-line options after extra `--`, and setting `TestingPlatformDotnetTestSupport` to true.

Reviewer: ```suggestion
To run MTP test applications in `dotnet test` VSTest mode, you should use `Microsoft.Testing.Platform.MSBuild`, pass MTP-specific command-line options after the extra --, and set `TestingPlatformDotnetTestSupport` to true.
```

2. VSTest-oriented command-line options are silently ignored.

> 3. Due to the above issues, we introduce a new `dotnet test` mode that is designed for MTP, and we encourage MTP users to move out of the VSTest `dotnet test` mode to the new mode with .NET 10 SDK.

Reviewer: ```suggestion
3. Due to these issues, we have introduced a new `dotnet test` mode specifically designed for MTP. We encourage MTP users to transition from the VSTest `dotnet test` mode to the new mode with the .NET 10 SDK.
```

> ## MTP mode of `dotnet test`

Reviewer: ```suggestion
## Microsoft.Testing.Platform (MTP) mode of `dotnet test`
```

> Due to the problems of running `dotnet test` with MTP in VSTest mode, we introduced a new mode of `dotnet test` in .NET 10 SDK, which is designed for MTP specifically.

Reviewer: ```suggestion
To address the issues encountered when running `dotnet test` with MTP in VSTest mode, we introduced a new mode in the .NET 10 SDK that is specifically designed for MTP.
```

> To opt-in this mode, you should add `dotnet.config` file in the root of the repository or the solution:

Reviewer: ```suggestion
To enable this mode, you should add a `dotnet.config` file to the root of the repository or solution.
```

```toml
[dotnet.test:runner]
name = "Microsoft.Testing.Platform"
```

> As this mode is designed for Microsoft.Testing.Platform, both `TestingPlatformDotnetTestSupport` and the extra `--` are no longer required.

Reviewer: ```suggestion
Since this mode is specifically designed for Microsoft.Testing.Platform, neither `TestingPlatformDotnetTestSupport` nor the additional `--` are required.
```

> [!IMPORTANT]
> This mode is only compatible for Microsoft.Testing.Platform 1.7.0 and later.
>
> [!IMPORTANT]

> > If you have a test project that supports VSTest and doesn't support MTP, an error is produced.

Reviewer: ```suggestion
> If your test project supports VSTest but does not support MTP, an error will be generated.
```
